### ISSUE TITLE
Shape Generation train jupyternotebook gives error for empty dataset

### ISSUE THREAD

**Issue by geekmangnu** [XinmiaoHuang/O-VITON-unofficial#2]

**Comment by XinmiaoHuang** [comment]
Hi geekmangnu:
&nbsp; &nbsp; &nbsp; &nbsp; Thanks for your email. For the errors you show, i suggest you check the directory of your dataset and make sure it is correct. It is not a complete project because i found it hard to reproduce the results of the paper.&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; Sorry that i couldn't maintain the code currently because I am now working on other project. But i will try it again when i am free.
&nbsp; &nbsp; &nbsp; &nbsp; Sorry again for those bugs. Wish you all the best.&nbsp;&nbsp;


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "XinmiaoHuang/O-VITON-unofficial"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2021年2月15日(星期一) 晚上11:13
收件人:&nbsp;"XinmiaoHuang/O-VITON-unofficial"<O-VITON-unofficial@noreply.github.com&gt;;
抄送:&nbsp;"Subscribed"<subscribed@noreply.github.com&gt;;
主题:&nbsp;[XinmiaoHuang/O-VITON-unofficial] Shape Generation train jupyternotebook gives error for empty dataset (#2)





 
Thank you for jupyternotebooks. I tried first shape genetaion train.ipynb
 I got following error.
 num_samples should be a positive integer value, but got num_samples=0
 from the code block below. It is because train_dataset length is 0.
 
`def lcm(a, b): return abs(a * b)/fractions.gcd(a, b) if a and b else 0
 
def main():
 opt = TrainOptions().parse(save=False)
 opt.checkpoints_dir = "/content/drive/My Drive/02_o-viton/source/shape_generation/checkpoints"
 opt.dataroot = "./dataset"
 opt.continue_train = True #이전 checkpoint부터 학습을 하고 싶으면 True.
 opt.gpu_ids = "0"
 opt.niter = 100
 opt.niter_decay = 0
 opt.batchSize = 4
 opt.print_freq = lcm(opt.print_freq, opt.batchSize) if opt.debug:     opt.display_freq = 1     opt.print_freq = 1     opt.niter = 1     opt.niter_decay = 0     opt.max_dataset_size = 10     opt.save_latest_freq = 1 #잠시만      print(f'opt.root: {opt.dataroot}') iter_path = os.path.join(opt.checkpoints_dir, opt.name, 'iter.txt') if opt.continue_train:     try:         start_epoch, epoch_iter = np.loadtxt(             iter_path, delimiter=',', dtype=int)     except:         start_epoch, epoch_iter = 1, 0     print('Resuming from epoch %d at iteration %d' % (start_epoch, epoch_iter)) else:     start_epoch, epoch_iter = 1, 0 # NEW DATALOADER augment = {} augment['1'] = transforms.Compose(     [         transforms.ToPILImage(),         transforms.RandomAffine(degrees=10, translate=(             0.1, 0.1), scale=(0.8, 1.2), shear=20),         transforms.ToTensor()])  # change to [C, H, W] augment['2'] = transforms.Compose(     [         transforms.ToTensor()])  # change to [C, H, W] train_dataset = RegularDataset(opt, augment) print('') print('') print(f'length of train_dataset:{len(train_dataset)}') print('') train_dataloader = DataLoader(train_dataset,                             batch_size=opt.batchSize,                             shuffle=True,                             num_workers=int(opt.nThreads),                             pin_memory=True) dataset_size = len(train_dataset) print('#training images = %d' % dataset_size) # FOR DEBUGGING print(" #Checking  the dimension and type of data") for key in train_dataset[0].keys():     try:         x = train_dataset[0][key]         print("name of the input and shape -- &gt; ", key, x.shape)         print("type,dtype,and min max -- &gt;", type(x),             x.dtype, torch.min(x), torch.max(x))     except Exception as e:         print("name of the input -- &gt; ", key, train_dataset[0][key])     print('----------------') dataset_size = len(train_dataset) print('#training images = %d' % dataset_size) # Initialize Networks model = create_model(opt) # Training Visualizer visualizer = Visualizer(opt) # Optimizers optimizer_G, optimizer_D = model.module.optimizer_G, model.module.optimizer_D # Train related additional details total_steps = (start_epoch-1) * dataset_size + epoch_iter display_delta = total_steps % opt.display_freq print_delta = total_steps % opt.print_freq save_delta = total_steps % opt.save_latest_freq for epoch in range(start_epoch, opt.niter + opt.niter_decay + 1):     epoch_start_time = time.time()     if epoch != start_epoch:         epoch_iter = epoch_iter % dataset_size     for i, data in enumerate(train_dataloader, start=epoch_iter):         if total_steps % opt.print_freq == print_delta:             iter_start_time = time.time()         total_steps += opt.batchSize         epoch_iter += opt.batchSize         # whether to collect output images         save_fake = total_steps % opt.display_freq == display_delta         ############## Forward Pass ######################         # Reference Purpose         # input_dict = {'seg_map': A_tensor, 'dense_map': dense_img, 'target': B_tensor, 'seg_map_path': A_path,         # 'target_path': A_path, 'densepose_path': dense_path }         # print( data['seg_mask'].shape)         losses, generated = model(             data['seg_map'], data['dense_map'], data['target'], data['seg_mask'], infer=save_fake)         # sum per device losses         losses = [torch.mean(x) if not isinstance(x, int)                 else x for x in losses]         loss_dict = dict(zip(model.module.loss_names, losses))         # calculate final loss scalar         loss_D = (loss_dict['D_fake'] + loss_dict['D_real']) * 0.5         loss_G = loss_dict['G_GAN'] + loss_dict.get('G_CE', 0)         ############### Backward Pass ####################         # update generator weights         optimizer_G.zero_grad()         loss_G.backward()         optimizer_G.step()         # update discriminator weights         optimizer_D.zero_grad()         loss_D.backward()         optimizer_D.step()         ############## Display results and errors ##########         # print out errors         if total_steps % opt.print_freq == print_delta:             errors = {k: v.data.item() if not isinstance(                 v, int) else v for k, v in loss_dict.items()}             t = (time.time() - iter_start_time) / opt.print_freq             visualizer.print_current_errors(epoch, epoch_iter, errors, t)             visualizer.plot_current_errors(errors, total_steps)             #call(["nvidia-smi", "--format=csv", "--query-gpu=memory.used,memory.free"])         # display output images         if save_fake:             visuals = OrderedDict([('input_label', util.tensor2label(data['seg_map'][0], opt.label_nc)),                                 ('synthesized_image', util.tensor2label(                                     generated.data[0], opt.label_nc)),                                 ('real_image', util.tensor2label(data['target'][0], opt.label_nc))])             visualizer.display_current_results(visuals, epoch, total_steps)         # save latest model         if total_steps % opt.save_latest_freq == save_delta:             print('saving the latest model (epoch %d, total_steps %d)' %                 (epoch, total_steps))             model.module.save('latest')             np.savetxt(iter_path, (epoch, epoch_iter), delimiter=',', fmt='%d')         if epoch_iter &gt;= dataset_size:             break     # end of epoch     iter_end_time = time.time()     print('End of epoch %d / %d \t Time Taken: %d sec' %         (epoch, opt.niter + opt.niter_decay, time.time() - epoch_start_time))     # save model for this epoch     if epoch % opt.save_epoch_freq == 0:         print('saving the model at the end of epoch %d, iters %d' %             (epoch, total_steps))         model.module.save('latest')         model.module.save(epoch)         np.savetxt(iter_path, (epoch+1, 0), delimiter=',', fmt='%d')     # linearly decay learning rate after certain iterations     if epoch &gt; opt.niter:         model.module.update_learning_rate()  
if name == 'main':
 # torch.multiprocessing.set_start_method('spawn')
 main()`
 
Before this it gaved another error. I passed it by creating
 ./dataset/train_label is not a valid directory
 Also it assumes /content/drive/My Drive/02_o-viton/source/shape_generation/checkpoints exists.
 Could you make it little bit more beginner friendly?
